OSDOCS#10947: Support for CCO in HCP (#77653)
openshift/openshift-docs: hosted_control_planes/hcp-authentication-authorization.adoc

The {product-title} control plane includes a built-in OAuth server. You can obtain OAuth access tokens to authenticate to the {product-title} API. After you create your hosted cluster, you can configure OAuth by specifying an identity provider.

include::modules/hcp-configuring-oauth.adoc[leveloffset=+1]

include::modules/hcp-configuring-oauth-console.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* To know more about supported identity providers, see xref:../authentication/understanding-identity-provider.adoc#understanding-identity-provider["Understanding identity provider configuration"] in _Authentication and authorization_.

[id="hcp-cco-aws-sts_{context}"]

> == Using the CCO in a hosted cluster on {aws-short}

Reviewer: ```suggestion
== Assigning components IAM roles by using the CCO in a hosted cluster on {aws-short}
```
Just a thought. Is the goal of this task to use the CCO in a hosted cluster, or is it to assign components IAM roles by using the CCO in a hosted cluster?

Author: As per the text `Assigning components IAM...` fits better. Addressed the suggestion.

You can assign components IAM roles that provide short-term, limited-privilege security credentials by using the Cloud Credential Operator (CCO) in hosted clusters on {aws-first}. By default, the CCO runs in a hosted control plane.

[NOTE]
====

> The CCO only supports a manual mode for hosted clusters on {aws-short}. By default, hosted clusters are configured in a manual mode. The management cluster might use modes other than manual.

Reviewer: ```suggestion
The CCO supports a manual mode only for hosted clusters on {aws-short}. By default, hosted clusters are configured in a manual mode. The management cluster might use modes other than manual.
```
I think "only" is modifying "a manual mode", so I moved "only" closer to that phrase. If "only" is actually modifying "supports", feel free to ignore this suggestion :)

Author: Ack'ed. Addressed

====

include::modules/hcp-cco-verify-aws-sts.adoc[leveloffset=+1]
include::modules/osdk-cco-aws-sts-enabling.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

> * xref:../operators/operator-reference.adoc#cloud-credential-operator_cluster-operators-ref[Cluster Operators reference page for the Cloud Credential Operator].

Reviewer: ```suggestion
* xref:../operators/operator-reference.adoc#cloud-credential-operator_cluster-operators-ref[Cluster Operators reference page for the Cloud Credential Operator]
```

The {product-title} control plane includes a built-in OAuth server. You can obtain OAuth access tokens to authenticate to the {product-title} API. After you create your hosted cluster, you can configure OAuth by specifying an identity provider.

include::modules/hcp-configuring-oauth.adoc[leveloffset=+1]

include::modules/hcp-configuring-oauth-console.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* To know more about supported identity providers, see xref:../authentication/understanding-identity-provider.adoc#understanding-identity-provider["Understanding identity provider configuration"] in _Authentication and authorization_.

> include::modules/cco-ccoctl-configuring.adoc[leveloffset=+3]

Reviewer 2: Might consider bumping this to a +2? Right now it looks like it's part of "Configuring the OAuth server", but IIUC it's not really part of that
```suggestion
include::modules/cco-ccoctl-configuring.adoc[leveloffset=+2]
```

Alternatively, you could make a subheading at the `==` level to put all the CCO stuff inside to more closely match the structure of these tasks in the install books. 

So, something like:
* Configuring the OAuth server for a hosted cluster by using the CLI
* Configuring the OAuth server for a hosted cluster by using the web console
* Configuring an {hcp} cluster to use the AWS Security Token Service
  * Configuring the Cloud Credential Operator utility
  * Creating AWS resources with the Cloud Credential Operator utility
    * Creating AWS resources with a single command
    * Creating AWS resources individually
  * Incorporating the Cloud Credential Operator utility manifests

This was a bit harder to show with a PR review diff, but I sketched it out [in my editor here](https://github.com/openshift/openshift-docs/compare/main...jeana-redhat:openshift-docs:cco-hcp?expand=1) if you want to see what I'm thinking :slightly_smiling_face: 

Local render TOC looks like this:
![image](https://github.com/openshift/openshift-docs/assets/61474374/7904efe6-933e-432d-95a4-7f4ac5b57b8a)

[id="hcp-sts-mode-ccoctl_{context}"]

> == Using Credential Operator utility for {hcp}

Reviewer 2: ```suggestion
== Creating {aws-short} resources with the Cloud Credential Operator utility
```

You have the following options when creating AWS resources:

> * You can use the `ccoctl aws create-all` command to create the AWS resources automatically. This is the quickest way to create the resources. See xref:../../../installing/installing_aws/ipi/installing-aws-network-customizations.adoc#cco-ccoctl-creating-at-once_installing-aws-network-customizations[Creating AWS resources with a single command].
> 
> * If you need to review the JSON files that the `ccoctl` tool creates before modifying AWS resources, or if the process the `ccoctl` tool uses to create AWS resources automatically does not meet the requirements of your organization, you can create the AWS resources individually. See xref:../../../installing/installing_aws/ipi/installing-aws-network-customizations.adoc#cco-ccoctl-creating-individually_installing-aws-network-customizations[Creating AWS resources individually].

Reviewer 2: For the links in these items, you probably want to point them to the instance of these modules on this page rather than off in IPI install

```suggestion
* You can use the `ccoctl aws create-all` command to create the AWS resources automatically. This is the quickest way to create the resources. See xref:../hosted_control_planes/hcp-authentication-authorization.adoc#cco-ccoctl-creating-at-once_hcp-authentication-authorization[Creating AWS resources with a single command].

* If you need to review the JSON files that the `ccoctl` tool creates before modifying AWS resources, or if the process the `ccoctl` tool uses to create AWS resources automatically does not meet the requirements of your organization, you can create the AWS resources individually. See xref:../hosted_control_planes/hcp-authentication-authorization.adoc#cco-ccoctl-creating-individually_hcp-authentication-authorization[Creating AWS resources individually].
```

include::modules/cco-ccoctl-creating-at-once.adoc[leveloffset=+4]

include::modules/cco-ccoctl-creating-individually.adoc[leveloffset=+4]

include::modules/cco-ccoctl-install-creating-manifests.adoc[leveloffset=+3]

> include::modules/cco-mode-sts.adoc[leveloffset=+1]

Reviewer 2: this doesn't seem to exist

[role="_additional-resources"]
.Additional resources

> * To know about the Cloud Credential Operator, see xref:../operators/operator-reference.adoc#cloud-credential-operator_cluster-operators-ref[Cluster Operators reference page for the Cloud Credential Operator].

Reviewer 2: Add'l resources should just have the link text :)
```suggestion
* xref:../operators/operator-reference.adoc#cloud-credential-operator_cluster-operators-ref[Cluster Operators reference page for the Cloud Credential Operator]
```

The {product-title} control plane includes a built-in OAuth server. You can obtain OAuth access tokens to authenticate to the {product-title} API. After you create your hosted cluster, you can configure OAuth by specifying an identity provider.

include::modules/hcp-configuring-oauth.adoc[leveloffset=+1]

include::modules/hcp-configuring-oauth-console.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* To know more about supported identity providers, see xref:../authentication/understanding-identity-provider.adoc#understanding-identity-provider["Understanding identity provider configuration"] in _Authentication and authorization_.

[id="hcp-ccoctl-aws-sts_{context}"]
== Configuring the CCO utility for a hosted cluster on {aws-short}

You can configure the Cloud Credential Operator (CCO) utility (`ccoctl`) on hosted clusters to assign components IAM roles that provide short-term, limited-privilege security credentials.

> To configure the CCO utility for a hosted cluster, your hosted cluster must use the {aws-short} {sts-short}.

Reviewer 2: For my education: are hosted clusters using AWS STS, or just manual CCO by itself? I ask because I wasn't aware that we could have AWS STS without the `ccoctl` :slightly_smiling_face:

> To configure the CCO utility for a hosted cluster, your hosted cluster must use the {aws-short} {sts-short}.

Reviewer 3: Let me try to explain in my comment next.
